File: packages/typescript/src/retry.ts
import { ProviderError } from "./errors.js";

export function jitter(base: number, cap: number, attempt: number): number {
  return Math.random() * Math.min(cap, base * Math.pow(2, attempt));
}

export async function retryAsync<T>(
  fn: () => Promise<T>,
  options: {
    maxAttempts?: number;
    baseDelay?: number;
    capDelay?: number;
    provider?: string;
  } = {},
): Promise<T> {
  const { maxAttempts = 3, baseDelay = 1000, capDelay = 30000, provider = "unknown" } = options;
  let lastErr: unknown;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (err) {
      lastErr = err;
      if (err instanceof ProviderError) {
        if (!err.retryable || attempt === maxAttempts - 1) throw err;
        const delay = err.retryAfter != null ? err.retryAfter * 1000 : jitter(baseDelay, capDelay, attempt);
        await sleep(delay);
      } else {
        if (attempt === maxAttempts - 1) {
          throw new ProviderError("transient", String(err), provider, true);
        }
        await sleep(jitter(baseDelay, capDelay, attempt));
      }
    }
  }
  throw lastErr;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export class CircuitBreaker {
  private failures: number[] = [];
  private openSince: number | null = null;

  constructor(
    private readonly failureThreshold: number = 5,
    private readonly windowMs: number = 60_000,
    private readonly recoveryMs: number = 30_000,
  ) {}

  recordFailure(): void {
    const now = Date.now();
    this.failures = this.failures.filter((t) => now - t < this.windowMs);
    this.failures.push(now);
    if (this.failures.length >= this.failureThreshold) {
      this.openSince = now;
    }
  }

  recordSuccess(): void {
    this.failures = [];
    this.openSince = null;
  }

  get isOpen(): boolean {
    if (this.openSince === null) return false;
    if (Date.now() - this.openSince > this.recoveryMs) {
      this.openSince = null;
      return false;
    }
    return true;
  }
}

File: packages/python/src/omniharness/_retry.py
from __future__ import annotations

import asyncio
import random
import time
from typing import Awaitable, Callable, TypeVar

from .errors import ProviderError

T = TypeVar("T")


async def retry_async(
    fn: Callable[[], Awaitable[T]],
    *,
    max_attempts: int = 3,
    base_delay: float = 1.0,
    cap_delay: float = 30.0,
    provider: str = "unknown",
) -> T:
    last_exc: Exception | None = None
    for attempt in range(max_attempts):
        try:
            return await fn()
        except ProviderError as e:
            last_exc = e
            if not e.retryable or attempt == max_attempts - 1:
                raise
            delay = e.retry_after or _jitter(base_delay, cap_delay, attempt)
            await asyncio.sleep(delay)
        except (OSError, asyncio.TimeoutError) as e:
            last_exc = e
            if attempt == max_attempts - 1:
                raise ProviderError("transient", str(e), provider, True) from e
            await asyncio.sleep(_jitter(base_delay, cap_delay, attempt))
    raise last_exc  # type: ignore[misc]


def _jitter(base: float, cap: float, attempt: int) -> float:
    return random.uniform(0, min(cap, base * (2**attempt)))


class CircuitBreaker:
    def __init__(
        self,
        failure_threshold: int = 5,
        window_seconds: float = 60.0,
        recovery_seconds: float = 30.0,
    ):
        self._threshold = failure_threshold
        self._window = window_seconds
        self._recovery = recovery_seconds
        self._failures: list[float] = []
        self._open_since: float | None = None

    def record_failure(self) -> None:
        now = time.monotonic()
        self._failures = [t for t in self._failures if now - t < self._window]
        self._failures.append(now)
        if len(self._failures) >= self._threshold:
            self._open_since = now

    def record_success(self) -> None:
        self._failures.clear()
        self._open_since = None

    @property
    def is_open(self) -> bool:
        if self._open_since is None:
            return False
        if time.monotonic() - self._open_since > self._recovery:
            self._open_since = None
            return False
        return True
